proposal/nested-map-functions.md

```diff
@@ -0,0 +1,322 @@
+
+* If `$keys` is empty, throw an error.
+
+* Let `key` be the first (or only) element in `$keys`
+
+* If `$map` does not have a key equal to `key`, return `null`.
+
+* Let `value` be the value assigned to `key` in `$map`
+
+* If there is more than one element in `$keys`:
+
+  * Let `keys` be all elements in `$keys` after the first element
+
+  * Call `get()` with `value` and expanded `keys` as arguments
+
+* Otherwise, return `value`
+
+### `has-key()`
+
+```
+has-key($map, $keys...)
+```
+
+* If `$map` is not a map, throw an error.
+
+* If `$keys` is empty, throw an error.
+
+* Let `key` be the first (or only) element in `$keys`
+
+* If `$map` does not have a key with the same name as `key`, return boolean
+  `false`.
+
+* If there is more than one element in `$keys`:
+
+  * Let `map` be the value assigned to `key` in `$map`
+
+  * Let `keys` be all but the first element in `$keys`
+
+  * Call `has-key()` with `map` and expanded `keys` as arguments
+
+* Otherwise, return boolean `true`
+
+### `set()`
+
+```
+set($map, $args...)
+```
+
+* If `$map` is not a map, throw an error.
+
+* Let `map` be an empty map.
+
+* Let `set-key` be the first item in arglist `$args`.
+
+* Let `remaining` be a slice of all the other items in arglist `$args`.
+
+* If there are no items in `remaining`, throw an error.
+
+* If there is more than one argument in `remaining`:
+
+  * If `$map` has a key `current-key` that is equal to `set-key`:
+
+    * Let `current` be the value of `current-key`.
+
+  * Otherwise:
+
+    * Let `current` be an empty map.
+
+  * Let `set-value` be the result of calling `set()` with `current` and expanded
+    `remaining` as arguments.
+
+    > This will error if `current` is not a map, but we stil have `remaining`
+
+* Otherwise:
+
+  * Let `set-value` be the only item in `remaining`
+
+* Return a copy of `map` with `set-key` set to `set-value`, overriding an
+  existing value for `set-key` if one exists.
+
+### `merge()`
+
+This proposal add a new overload to the existing `merge()` function:
+
+```
+merge($map1, $args...)
+```
+
+* If `$args` is empty, throw an error
+
+* Let `map2` be the last item in arglist `$args`
+
+* If either `$map1` or `map2` is not a map, throw an error.
+
+* If arglist `$args` has more than one item:
+
+  * Let `get-keys` be a slice of all except the last item in `args`.
+
+  * Let `sub` be the result of calling `get()` with `sub` and expanded `get-keys`
+    as arguments.
+
+  * Let `sub-merged` be the result of calling `merge()` with `sub` and `map2`
+    as argumets.
+
+  * Let `set-args` be the result of appending `sub-merged` to the list `keys`
+
+  * Return the result of calling `set()` with `$map1` and expanded `set-args`
+    as arguments.
+
+* Otherwise:
+
+  * Let `map` be a copy of `$map1`.
+
+  * For each `key`, `value` in `map2`:
+
+    * Replace `map` with the result of calling `set()` with `map`, `key`, and
+      `value` as arguments.
+
+  * Return `map`.
```

Comment:
Wasn't sure how much this current two-map behavior needs to be spelled out. I maybe landed on a strange middle-ground?

Reply:
I think this is reasonable, but I might be more explicit rather than relying on calling other functions. Something like:

> * For each `key`, `value` in `map2`:
>
>   * Set `key` to `value` in `map`, replacing an existing value for `key` if one exists.

Same with the call to `get()` above.